h/atom_feed.py

```diff
@@ -0,0 +1,173 @@
+        "{text}".format(
+            selection=cgi.escape(get_selection(annotation)),
+            text=cgi.escape(annotation["text"])))
+
+    entry["links"] = []
+
+    entry["links"].append({"rel": "alternate", "type": "text/html",
+                           "href": annotation_url(annotation)})
+
+    if annotation_api_url:
+        entry["links"].append({"rel": "alternate", "type": "application/json",
+                               "href": annotation_api_url(annotation)})
+
+    return entry
+
+
+def _feed_from_annotations(
+        annotations, atom_url, annotation_url, annotation_api_url=None,
+        html_url=None, title=None, subtitle=None):
+    """Return an Atom feed for the given list of annotations.
+
+    This returns a logical representation of an Atom feed as a Python dict
+    containing all of the data that a template would need to render the feed
+    to XML (including a list of dicts for the feed's entries).
+
+    """
+    links = [{"rel": "self", "type": "application/atom+xml", "href": atom_url}]
+
+    if html_url:
+        links.append(
+            {"rel": "alternate", "type": "text/html", "href": html_url})
+
+    entries = [
+        _feed_entry_from_annotation(a, annotation_url, annotation_api_url)
+        for a in annotations]
+
+    feed = {
+        "id": atom_url,
+        "title": title or _("Hypothesis Stream"),
+        "subtitle": subtitle or _("The Web. Annotated"),
+        "entries": entries,
+        "links": links
+    }
+
+    if annotations:
+        feed["updated"] = annotations[0]["updated"]
+
+    return feed
+
+
+def render(
+        request, annotations, atom_url, html_url, title=None, subtitle=None):
+    """Return an Atom feed of the given list of annotations.
+
+    :param annotations: A list of annotations from the API
+    :type annotations: list of dicts
+
+    :param atom_url: The URL where this Atom feed will be hosted
+        (the feed will contain a link to this URL)
+    :type atom_url: unicode
+
+    :param html_url: The URL to the HTML page that this Atom feed is a feed
+        for (the feed will contain a link to this URL).
+    :type html_url: unicode
+
+    :param title: The title of the feed
+    :type title: unicode
+
+    :param subtitle: The subtitle of the feed
+    :type subtitle: unicode
+
+    :returns: An Atom feed as an XML string
+    :rtype: unicode
+
+    """
+    def annotation_url(annotation):
+        """Return the HTML permalink URL for the given annotation."""
+        return request.resource_url(request.root, "a", annotation["id"])
+
+    def annotation_api_url(annotation):
+        """Return the JSON API URL for the given annotation."""
+        return request.resource_url(request.root, "api", "annotation",
+                                    annotation["id"])
+
+    feed = _feed_from_annotations(
+        annotations=annotations, atom_url=atom_url,
+        annotation_url=annotation_url, annotation_api_url=annotation_api_url,
```

Comment:
Should we just be using named routes to make this clearer?

`request.route_url('annotation', id='...')`

I very much like the theory of Pyramid's traversal-based routing, but it makes most sense when there's a deep tree of resources and particularly where users can create trees of resources. For our top-level resources maybe we just shouldn't bother if it's making things more difficult.

Also, maybe we should do content negotiation on the annotation resource rather than having to construct a rel alternate to the API server.